homeassistant/components/kraken/sensor.py

```diff
@@ -0,0 +1,212 @@
+            )
+        }
+
+        for tracked_asset_pair in config_entry.options[CONF_TRACKED_ASSET_PAIRS]:
+            # Only create new devices
+            if create_device_name(tracked_asset_pair) in existing_devices.keys():
+                existing_devices.pop(create_device_name(tracked_asset_pair))
+            else:
+                sensors = []
+                for sensor_type in SENSOR_TYPES:
+                    sensors.append(
+                        KrakenSensor(
+                            hass.data[DOMAIN],
+                            tracked_asset_pair,
+                            sensor_type,
+                        )
+                    )
+                async_add_entities(sensors, True)
+
+        # Remove devices for asset pairs which are no longer tracked
+        for device_id in existing_devices.values():
+            device_registry.async_remove_device(device_id)
+
+    await async_update_sensors(hass, config_entry)
+
+    hass.data[DOMAIN].unsub_listeners.append(
+        async_dispatcher_connect(
+            hass,
+            DISPATCH_CONFIG_UPDATED,
+            async_update_sensors,
+        )
+    )
+
+
+class KrakenSensor(CoordinatorEntity):
+    """Define a Kraken sensor."""
+
+    def __init__(
+        self,
+        kraken_data: KrakenData,
+        tracked_asset_pair: str,
+        sensor_type: str,
+    ) -> None:
+        """Initialize."""
+        super().__init__(kraken_data.coordinator)
+        self.tracked_asset_pair_wsname = kraken_data.tradable_asset_pairs[
+            tracked_asset_pair
+        ]
+        self._source_asset = tracked_asset_pair.split("/")[0]
+        self._target_asset = tracked_asset_pair.split("/")[1]
+        self._sensor_type = sensor_type
+        self._unit_of_measurement = self._target_asset
+        self._device_name = f"{self._source_asset} {self._target_asset}"
+        self._name = "_".join(
+            [
+                tracked_asset_pair.split("/")[0],
+                tracked_asset_pair.split("/")[1],
+                sensor_type,
+            ]
+        )
+        self._received_data_at_least_once = False
+        self._available = True
+
+    @property
+    def name(self):
+        """Return the name."""
+        return self._name
+
+    @property
+    def unique_id(self):
+        """Set unique_id for sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state."""
+        try:
+            state = self._try_get_state()
+            self._received_data_at_least_once = True  # Received data at least one time.
+            return state
+        except TypeError:
+            if self._received_data_at_least_once:
+                if self._available:
+                    _LOGGER.warning(
+                        "Asset Pair %s is no longer available",
+                        self._device_name,
+                    )
+                    self._available = False
```

Comment:
Move this logic to `_handle_coordinator_update` which you inherit from `CoordinatorEntity`.

Reply:
Done